lib/src/iterable_extensions.dart

```diff
@@ -0,0 +1,519 @@
+    if (!iterator.moveNext()) {
+      return;
+    }
+    var index = 1;
+    var chunk = [iterator.current];
+    while (iterator.moveNext()) {
+      var element = iterator.current;
+      if (test(index++, element)) {
+        yield chunk;
+        chunk = [element];
+      }
+    }
+    yield chunk;
+  }
+
+  /// Splits the elements into chunks after some elements and indices.
+  ///
+  /// Each element and index is checked using [test]
+  /// for whether it should end the current chunk.
+  /// If so, the elements since the previous chunk-ending element
+  /// are emitted as a list.
+  /// Any final elements are emitted at the end, whether the last
+  /// element should be split after or not.
+  ///
+  /// Example:
+  /// ```dart
+  /// var parts = [1, 0, 2, 1, 5, 7, 6, 8, 9].splitAfterIndexed((i, v) => i < v);
+  /// print(parts); // ([1, 0], [2, 1], [5, 7, 6], [8, 9])
+  /// ```
+  Iterable<List<T>> splitAfterIndexed(bool test(int index, T element)) sync* {
+    var index = 0;
+    List<T> chunk;
+    for (var element in this) {
+      (chunk ??= []).add(element);
+      if (test(index++, element)) {
+        yield chunk;
+        chunk = null;
+      }
+    }
+    if (chunk != null) yield chunk;
+  }
+
+  /// Splits the elements into chunks between some elements and indices.
+  ///
+  /// Each pair of adjacent elements and the index of the latter are
+  /// checked using [test] for whether a chunk should end between them.
+  /// If so, the elements since the previous chunk-splitting elements
+  /// are emitted as a list.
+  /// Any final elements are emitted at the end.
+  ///
+  /// Example:
+  /// ```dart
+  /// var parts = [1, 0, 2, 1, 5, 7, 6, 8, 9].splitBetweenIndexed((i, v1, v2) => v1 > v2);
+  /// print(parts); // ([1], [0, 2], [1, 5, 7], [6, 8, 9])
+  /// ```
+  Iterable<List<T>> splitBetweenIndexed(
+      bool test(int index, T first, T second)) sync* {
+    var iterator = this.iterator;
+    if (!iterator.moveNext()) return;
+    var previous = iterator.current;
+    var chunk = <T>[previous];
+    var index = 1;
+    while (iterator.moveNext()) {
+      var element = iterator.current;
+      if (test(index++, previous, element)) {
+        yield chunk;
+        chunk = [];
+      }
+      chunk.add(element);
+      previous = element;
+    }
+    yield chunk;
+  }
+}
+
+/// Extensions that apply to iterables of nullable elements.
+extension IterableNullableExtension<T extends Object> on Iterable<T /*?*/ > {
+  Iterable<T> whereNotNull() sync* {
+    for (var element in this) {
+      if (element != null) yield element;
+    }
+  }
+}
+
+/// Extensions that apply to iterables of numbers.
+extension IterableNumberExtension<T extends num> on Iterable<T> {
+  /// The sum of the elements.
+  ///
+  /// The sum is zero if the iterable is empty.
+  T sum() {
+    var result = (0.5 is T ? 0.0 : 0) as T;
+    for (var value in this) {
+      result += value;
+    }
+    return result;
+  }
+
+  /// The average of the elements.
```

Comment:
We need to call out that this is the mean, not the median. Or we could name it `mean` or `arithmeticMean` and avoid the ambiguous `average` term altogether.